### ISSUE TITLE
Negative type constraint

### ISSUE THREAD

**Issue by mleonhard** [dart-lang/sdk#46114]
Dart currently lets us write
```dart
T someFunc<T extends Object>(T Function() f) {...}
```

I repeatedly make the mistake of passing `Future<T> Function()` to a function that expects `T Function()`.  These mistakes often cause runtime errors.  I wish to prevent such errors with a type constraint that means "T does not extend Future".  Proposed syntax:
```dart
T someFunc<T extends Object + !Future>(T Function() f) {...}
```

A linter warning would also solve the problem.

**Comment by tatumizer** [comment]
The more realistic syntax for this constraint would be: `@nonFuture` annotation (to be defined in meta package)

**Comment by srawlins** [comment]
Agreed, this is more likely to get resolved via a linter rule and maybe an annotation.

**Comment by bwilkerson** [comment]
One proposal for an annotation: https://github.com/dart-lang/sdk/issues/61105
